packages/ui/src/Density/index.tsx

```diff
@@ -1,11 +1,29 @@
 import { ReactNode } from 'react';
 import { Density as TDensity, DensityContext } from '../utils/density';
 
-export type DensityProps<SelectedDensity extends TDensity> = {
+/**
+ * Utility interface to be used below to ensure that only one density type is used at a time.
+ */
+interface NeverDensityTypes {
+  sparse?: never;
+  medium?: never;
+  compact?: never;
+}
+
+export type DensityPropType =
+  | (Omit<NeverDensityTypes, 'sparse'> & {
+      sparse: true;
+    })
+  | (Omit<NeverDensityTypes, 'medium'> & {
+      medium: true;
+    })
+  | (Omit<NeverDensityTypes, 'compact'> & {
+      compact: true;
+    });
+
+export type DensityProps = DensityPropType & {
   children?: ReactNode;
-} & (SelectedDensity extends 'sparse'
-  ? { sparse: true; compact?: never }
-  : { compact: true; sparse?: never });
+};
```

Comment:
question: can't we do this?

```ts
export type Density = 'sparse' | 'medium' | 'compact';

export type DensityProps = {
  density: Density;
  children?: ReactNode;
};
```

Reply:
though if we specifically need those names as a field, we could also do this:

```ts
export type DensityPropType =
  | { sparse: true; medium?: never; compact?: never }
  | { medium: true; sparse?: never; compact?: never }
  | { compact: true; sparse?: never; medium?: never };

export type DensityProps = DensityPropType & {
  children?: ReactNode;
};
```

Reply:
Density and Text components initially had this style of props that is more neat: `<Density compact/>`, `<Text body/>`. I personally don't mind the union type but it is already a bit too late to change – so many components depend on a shorter syntax.

Changed the types to the version from your second comment